src/validation/statement.rs

```diff
@@ -822,20 +822,70 @@ fn validate_assignment<T: AnnotationMap>(
 }
 
-pub(crate) fn validate_enum_variant_assignment(
+pub(crate) fn validate_enum_variant_assignment<T: AnnotationMap>(
+    context: &ValidationContext<T>,
     validator: &mut Validator,
-    left: &DataTypeInformation,
-    right: &DataTypeInformation,
     qualified_name: &str,
-    location: SourceLocation,
+    left_dt: &DataType,
+    right: &AstNode,
 ) {
-    if left.is_enum() && left.get_name() != right.get_name() {
-        validator.push_diagnostic(
-            Diagnostic::error(format!("Assigned value is not a variant of {qualified_name}"))
-                .with_error_code("E039")
-                .with_location(location),
-        )
+    if !left_dt.is_enum() {
+        return;
     }
+
+    let right_dt = context.annotations.get_type_or_void(right, context.index);
+
+    // For it to be a valid enum assignment, the right-hand side must yield a const-expr value
+    // (i.e. literal integer or some enum variant) and the left-hand side (which is an enum) must have that
+    // const-expr value as a variant (e.g. the const-expr must be 1 or 2 for `Status : (idle := 1, running := 2)`)
+    let Some(value_rhs) = get_literal_int_or_const_expr_value(right, context) else {
+        // ...however function calls for example are no const-expr hence only report if datatypes also differ
+        if left_dt.get_name() != right_dt.get_name() {
+            validator.push_diagnostic(
+                Diagnostic::error(format!(
+                    "Value evaluated at run-time, use an enum variant from `{}`",
+                    get_datatype_name_or_slice(validator.context, left_dt)
+                ))
+                .with_location(right.get_location())
+                .with_error_code("E091")
+            );
+        }
+
+        return;
+    };
+
+    let Some(variable) = context.index.find_fully_qualified_variable(qualified_name) else { return };
+    let variants = context.index.get_enum_variants(variable);
+
+    match variants.iter().find(|(_, _, value_lhs)| *value_lhs == value_rhs) {
+        Some((name, variant, _)) => {
+            if left_dt.get_name() != right_dt.get_name() {
+                // TODO(volsa) This sucks, locally defined enums have mangled names instead of normalized ones
+                let qualified_enum_name =
+                    if name.starts_with("__") { variant.to_string() } else { format!("{name}.{variant}") };
+
+                validator.push_diagnostic(
+                    Diagnostic::info(format!(
+                        "Consider replacing `{}` with `{qualified_enum_name}`",
+                        validator.context.slice(&right.location)
+                    ))
+                    .with_location(right.get_location())
+                    .with_secondary_location(left_dt.location.clone()),
+                );
+            }
+        }
+        None => {
+            validator.push_diagnostic(
+                Diagnostic::error(format!(
+                    "`{}` is an invalid value for enum `{}`",
```

Comment:
On the assumption that aliasing enum-types is allowed, this might fail on aliased types.
```
TYPE State : (Idle := 0, Working := 100); END_TYPE
TYPE StateAlias: State; END_TYPE

FUNCTION FOO : State
    FOO := Idle;
END_FUNCTION

FUNCTION BAR : StateAlias
    BAR := Idle; 
           ^^^^ `Idle` is an invalid value for enum `State`
END_FUNCTION
```

But I'm honestly not sure how aliasing enums should behave, since changing the return to `StateAlias.Idle` will result in an unresolved reference error.

Reply:
To add to this, I don't think this should block merging, just something to investigate down the line.

Reply:
I really hope no sane person aliases an enum (what would be the use case for this in the first place?). That being said, really good test-case catch.

> But I'm honestly not sure how aliasing enums should behave

Same, maybe @ghaith has some opinion on it?

Reply:
Aliasing an enum would indeed be strange. Let's agree that this would trigger a false positive for now, maybe even document that it does in a test so that when it's an issue we know why it's there.